File: dbassistant/layers/GlossaryConversionLayer.py
import re
from typing import List
from dbassistant.interfaces.ILayer import ILayer
from dbassistant.layers.LayerCargo import LayerCargo
from dbassistant.tokens.Tokenizer import Tokenizer


class GlossaryConversionLayer(ILayer):
    def __init__(self, tokenizer:Tokenizer):
        self.tokenizer = tokenizer

    def process(self, cargo: LayerCargo) -> LayerCargo:
        input_string = cargo.take()
        assert isinstance(input_string, str)
        
        looking_for = self.tokenizer.glossaryTokenHandler.keyMap.keys()
        for key in looking_for:
            input_string = re.sub('\\b' + key + '\\b', key.replace(" ", "_"), input_string)

        cargo.put(input_string)
        return cargo

File: dbassistant/layers/GlossaryConversionLayer.ts
import ILayer from '../interfaces/ILayer';
import LayerCargo from '../layers/LayerCargo';
import Tokenizer from '../tokens/Tokenizer';

export default class GlossaryConversionLayer implements ILayer {
    private tokenizer: Tokenizer;

    constructor(tokenizer: Tokenizer) {
        this.tokenizer = tokenizer;
    }

    process(cargo: LayerCargo): LayerCargo {
        let input_string = cargo.take();
        if (typeof input_string !== 'string') {
            throw new Error('Input cargo should be a string.');
        }

        const looking_for = Object.keys(this.tokenizer.glossaryTokenHandler.keyMap);
        for (const key of looking_for) {
            const pattern = '\\b' + key.replace('+', '\\+') + '\\b';
            const replacement = key.replace(' ', '_');
            input_string = input_string.replace(new RegExp(pattern, 'g'), replacement);
        }

        cargo.put(input_string);
        return cargo;
    }
}
